fix: Hiding widgets / links from unauthorized users on welcome screen.

TommyJackson85 opened the pull request:
## Related Issue
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->
Closes #1074

## Your solution
<!--- Please describe your solution, have you encountered any issues along the way? -->
Attempting to hide ```welcomeScreenWidget```s from unauthorised users, as they should not have access to if they're not authorised to use them.\
Using ```app-form-builder``` as a first attempt.

My initial attempt was to do it from each welcomeScreenWidget but I only could manage to hide the link button. Here is what it looked like:

<img width="1294" alt="Screenshot 2020-06-28 at 13 19 07" src="https://user-images.githubusercontent.com/22279028/85947343-fdc85980-b941-11ea-926f-e564977b9ca5.png">

So I then applied SecureView to each mapped out Widget plugin from ```app-plugin-admin-welcome-screen/src/components/Welcome.tsx``` file, while passing in a scopes key from the app-form-builder welcomeScreenWidget plugin.
From both unauthorized and authorized accounts, here are the results as it appears to work:

<img width="977" alt="Screenshot 2020-06-28 at 13 49 00" src="https://user-images.githubusercontent.com/22279028/85948167-3e76a180-b947-11ea-9972-7b2f2b810923.png">

<img width="990" alt="Screenshot 2020-06-28 at 13 48 00" src="https://user-images.githubusercontent.com/22279028/85948170-40d8fb80-b947-11ea-90d9-255bbd15b0c2.png">

@EmilK15 what do you think? Should I keep the Secure View from the Welcome.tsx file instead and just pass in Secure View scopes from welcomeScreenWidget? I am still yet to do this to other widgets. Just doing one as a first attempt.

## How Has This Been Tested?
I am testing visually and manually through both authorized and unauthorized accounts
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

TommyJackson85 commented:
@EmilK15  I updated my initial post to try something different. :)

Pavel910 commented on packages/app-form-builder/src/admin/plugins/welcomeScreenWidget.tsx:
```diff
@@ -19,15 +22,17 @@ const buttonStyle = css({
 const plugin: AdminWelcomeScreenWidgetPlugin = {
     type: "admin-welcome-screen-widget",
     name: "admin-welcome-screen-widget-form-builder",
+    scopes: ROLE_FORMS_EDITOR,
```
I think this is an acceptable approach 👍

Pavel910 commented on packages/app-form-builder/src/admin/plugins/welcomeScreenWidget.tsx:
```diff
@@ -19,15 +22,17 @@ const buttonStyle = css({
-            <Link to="/forms" className={linkStyle}>
-                <ButtonSecondary className={buttonStyle}>Create a new Form</ButtonSecondary>
-            </Link>
+            <SecureView scopes={ROLE_FORMS_EDITOR}>
```
You can safely remove this `SecureView`.

Pavel910 commented on packages/app-plugin-admin-welcome-screen/src/components/Welcome.tsx:
```diff
@@ -124,23 +125,25 @@ const Welcome = () => {
-                                                </div>
-                                            </Elevation>
-                                        </Widget>
+                                        <SecureView scopes={pl.scopes}>
```
This looks good to me. Let's go with this approach.

TommyJackson85 commented:
Yeah for sure, I was considering considering converting it to a function but it could of been
set as an Object for a reason. each object is mapped out as jsx using map from Welcome.tsx.

I agree with you on removing SecureView on each Widget too. I had no plans to keep applying SecureView to the buttons just the widgets. Just left it there to show what I tried.
Better to just pass each scopes from them instead. I will work on it later.

Sent from my iPhone

> On 28 Jun 2020, at 14:21, Pavel Denisjuk <notifications@github.com> wrote:
> 
> ﻿
> @Pavel910 requested changes on this pull request.
> 
> @TommyJackson85 this approach with defining scopes in the widget plugin is ok. Let's go that way. The only thing to remove is that SecureView from each plugin itself.
> 
> Another approach to the whole problem would be to convert widget on each plugin into a function, which would receive a Widget component, and then you'd only need to render it. That way the control would be in each individual plugin.
> 
> Hard to tell which one is better at this point.
> 
> In packages/app-form-builder/src/admin/plugins/welcomeScreenWidget.tsx:
> 
> > @@ -19,15 +22,17 @@ const buttonStyle = css({
>  const plugin: AdminWelcomeScreenWidgetPlugin = {
>      type: "admin-welcome-screen-widget",
>      name: "admin-welcome-screen-widget-form-builder",
> +    scopes: ROLE_FORMS_EDITOR,
> I think this is an acceptable approach 👍
> 
> In packages/app-form-builder/src/admin/plugins/welcomeScreenWidget.tsx:
> 
> >      widget: {
>          cta: (
> -            <Link to="/forms" className={linkStyle}>
> -                <ButtonSecondary className={buttonStyle}>Create a new Form</ButtonSecondary>
> -            </Link>
> +            <SecureView scopes={ROLE_FORMS_EDITOR}>
> You can safely remove this SecureView.
> 
> In packages/app-plugin-admin-welcome-screen/src/components/Welcome.tsx:
> 
> > -                                                <Typography use={"headline6"}>
> -                                                    <p className={widgetTitleStyle}>
> -                                                        {pl.widget.title}
> -                                                    </p>
> -                                                </Typography>
> -                                                <Typography use={"body1"}>
> -                                                    <p className={widgetDescriptionStyle}>
> -                                                        {pl.widget.description}
> -                                                    </p>
> -                                                </Typography>
> -                                                <div className={widgetButtonStyle}>
> -                                                    {pl.widget.cta}
> -                                                </div>
> -                                            </Elevation>
> -                                        </Widget>
> +                                        <SecureView scopes={pl.scopes}>
> This looks good to me. Let's go with this approach.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

TommyJackson85 commented:
@Pavel910  @EmilK15  I removed the unnecessary SecureViews from the widget and added the remaining scopes key values to the other displayed widgets. Now on unauthorised user accounts it looks like this:

<img width="709" alt="Screenshot 2020-06-28 at 21 02 59" src="https://user-images.githubusercontent.com/22279028/85957179-26247800-b983-11ea-99cf-2a9e319b8702.png">

I think for user accounts who have no permissions set this above message should be removed or edited to something more welcoming and applicable because right now they don't have any actions to perform so it's misleading. 
Maybe it should be for a new issue though.

On a side note, do you where else in the project to test the code? Like automated tests? If they exist I could see they cover the code I wrote..

doitadrian commented:
@TommyJackson85, for Cypress tests, please check the https://github.com/webiny/webiny-js/blob/master/docs/CONTRIBUTING.md#cypress-tests. 

Let us know if you'll have additional questions.

Pavel910 commented:
@TommyJackson85 regarding an empty Welcome screen - let's do it this way: when you get all plugins, you have an array of scopes required by each particular widget. If the user doesn't have access to any of the widgets, we can print a message to contact the administrator. I think this will be enough for now. 

You can check plugin scopes programmatically, by using `import { hasScopes } from "@webiny/app-security";` and render the message accordingly.

I really like how you try to handle all the different edge cases 🚀 🍻 👏

TommyJackson85 commented:
> @TommyJackson85 regarding an empty Welcome screen - let's do it this way: when you get all plugins, you have an array of scopes required by each particular widget. If the user doesn't have access to any of the widgets, we can print a message to contact the administrator. I think this will be enough for now.
> 
> You can check plugin scopes programmatically, by using `import { hasScopes } from "@webiny/app-security";` and render the message accordingly.
> 
> I really like how you try to handle all the different edge cases 🚀 🍻 👏

@Pavel910 I think i managed to use ```hasScopes``` correctly. :) Can you double check my recent commits? It displays a different messaged for users with no authorization. I manually tested it and it seems to work.

@doitadrian thanks! :) Will check it over the week.

TommyJackson85 commented:
@Pavel910 my apologies, I didn't get hasScopes it right yet.

Pavel910 commented:
Hey @TommyJackson85, sorry for the delay. This is what I think will solve the problem:

```tsx
    // Check if user has at least 1 scope
    const canSeeAnything = getPlugins<AdminWelcomeScreenWidgetPlugin>(
        "admin-welcome-screen-widget"
    ).some(pl => hasScopes(pl.scopes, { forceBoolean: true }));
```

We need to see if a user can see at least 1 of the widgets.  Each `Widget` plugin has `scopes` in form of an array. `Array.some` will tell us if at least 1 element passes the test. And `hasScopes` will process an array of scopes defined in the widget plugin, and tell you whether the use has all of the required scopes. It works in the `AND` manner, meaning that the user must have ALL requested scopes for `hasScopes` to return `true`. 

Let me know if this helps!

TommyJackson85 commented:
i thaught it would be fixed as all checks have passed but i am getting this error or running the app locally. The Webiny loading icon appears then this error appears:

<img width="1026" alt="Screenshot 2020-07-02 at 13 50 31" src="https://user-images.githubusercontent.com/22279028/86361108-41390580-bc6b-11ea-8b04-fe108df371f9.png">
<img width="1005" alt="Screenshot 2020-07-02 at 13 50 16" src="https://user-images.githubusercontent.com/22279028/86361112-4302c900-bc6b-11ea-9b44-b4ff1f791e28.png">

Pavel910 commented on packages/app-plugin-admin-welcome-screen/src/components/Welcome.tsx:
```diff
@@ -112,10 +122,18 @@ const Welcome = () => {
-                                        To get started - pick one of the actions below:
-                                    </p>
+                                <Typography use={"headline6"}>      
+                                    <h6>{getPlugins<AdminWelcomeScreenWidgetPlugin>("admin-welcome-screen-widget")}</h6>
```
@TommyJackson85 this line here is causing your React error. You're trying to render an array of objects. Not sure what you wanted to render here, but this is the place to fix.

TommyJackson85 commented:
Oh I see. I was only doing that to test to see what it prints 😬was gonna delete later.

Sent from my iPhone

> On 2 Jul 2020, at 13:58, Pavel Denisjuk <notifications@github.com> wrote:
> 
> ﻿
> @Pavel910 commented on this pull request.
> 
> In packages/app-plugin-admin-welcome-screen/src/components/Welcome.tsx:
> 
> > @@ -112,10 +122,18 @@ const Welcome = () => {
>                      <SimpleFormContent>
>                          <ContentTheme>
>                              <Cell span={12}>
> -                                <Typography use={"headline6"}>
> -                                    <p className={pGetStartedStyle}>
> -                                        To get started - pick one of the actions below:
> -                                    </p>
> +                                <Typography use={"headline6"}>      
> +                                    <h6>{getPlugins<AdminWelcomeScreenWidgetPlugin>("admin-welcome-screen-widget")}</h6>
> @TommyJackson85 this line here is causing your React error. You're trying to render an array of objects. Not sure what you wanted to render here, but this is the place to fix.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

TommyJackson85 commented:
How do I grant permission for a user to use each widget? ( i.e. from an administrative account I want to grant an unauthorized account permissions for each widget access). This is just so I can do further manual testing on an account to get authorization for one or two widgets.

So far from and administrator account, I can only give a user these permissions:
<img width="519" alt="Screenshot 2020-07-02 at 21 19 29" src="https://user-images.githubusercontent.com/22279028/86406978-0905e700-bcac-11ea-8713-12151dc3331e.png">

Anyway, I will give the Cyprus tests a shot, unless you want to merge these changes as they are.

Pavel910 commented:
@TommyJackson85 roles contain scopes. So you need to create a Role (or several of them) with a specific set of scopes. This way we allow administrators to structure their permissions however they like, and we only care the scopes at the code level.

TommyJackson85 commented:
Ok I’ll try it later, I need to delete a comment I left too.

Sent from my iPhone

> On 3 Jul 2020, at 09:06, Pavel Denisjuk <notifications@github.com> wrote:
> 
> ﻿
> @TommyJackson85 roles contain scopes. So you need to create a Role (or several of them) with a specific set of scopes. This way we allow administrators to structure their permissions however they like, and we only care the scopes at the code level.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

TommyJackson85 commented:
I think I managed to grant permissions for page building, but they only seem to appear in the side nav bar :( 
<img width="1014" alt="Screenshot 2020-07-04 at 13 01 43" src="https://user-images.githubusercontent.com/22279028/86512077-ae2cd680-bdf6-11ea-959b-9b1589ce94b6.png">

Pavel910 commented:
@TommyJackson85 that's because the scope in widget plugins are not really related to what you have assigned to the user. In the Page Builder widget you have `scopes: ["pb:settings"]`, should be `pb:page:crud`, and so on for all the plugins.

TommyJackson85 commented:
> @TommyJackson85 that's because the scope in widget plugins are not really related to what you have assigned to the user. In the Page Builder widget you have `scopes: ["pb:settings"]`, should be `pb:page:crud`, and so on for all the plugins.

That is fair enough. So 'pb:settings' has nothing to do with the widget plugins? Should I have given different scopes for the widgets instead of 'settings' scopes? Whould I have given a list of all 'pb scopes maybe'?

TommyJackson85 commented:
@Pavel910 would I be better to apply all potential scope types to the the scopes array including both ```pb:settings``` and ```pb:page:crud``` for example?

Pavel910 commented:
@TommyJackson85 I think `pb:page:crud` makes the most sense, because the widget is linking to Pages List, which is a module for creating pages. `pb:settings` is not relevant here. Also since `hasScopes` works in an `AND` fashion, you need to be careful not to include scopes that may not be present in user's token.

TommyJackson85 commented:
Ok so I will just use  page:crud   scopes for now so. thanks for the advice!

TommyJackson85 commented:
**These are the scopes I found, as you see there is nothing for CMS**

<img width="365" alt="Screenshot 2020-07-04 at 19 54 52" src="https://user-images.githubusercontent.com/22279028/86519186-590db680-be30-11ea-9758-ff1094dfc850.png">
<img width="318" alt="Screenshot 2020-07-04 at 19 54 40" src="https://user-images.githubusercontent.com/22279028/86519187-5a3ee380-be30-11ea-890c-d09a7265af3c.png">
<img width="412" alt="Screenshot 2020-07-04 at 19 53 57" src="https://user-images.githubusercontent.com/22279028/86519188-5ad77a00-be30-11ea-8092-6455d9e6aac0.png">
<img width="432" alt="Screenshot 2020-07-04 at 19 52 33" **src="https://user-images.githubusercontent.com/22279028/86519223-a722ba00-be30-11ea-9cb3-a7bf32a8ec73.png">**
<img width="289" alt="Screenshot 2020-07-04 at 19 50 35" src="https://user-images.githubusercontent.com/22279028/86519226-a853e700-be30-11ea-9099-c36124324573.png">
<img width="286" alt="Screenshot 2020-07-04 at 19 50 20" src="https://user-images.githubusercontent.com/22279028/86519227-a8ec7d80-be30-11ea-9d27-0cd0554c420e.png">

Pavel910 commented:
@TommyJackson85 CMS scopes are in progress in another PR: https://github.com/webiny/webiny-js/pull/1088.
You can safely ignore that one for now, simply set the scopes to `[]` to let that widget always be visible, and we'll come back to it later.

TommyJackson85 commented:
> @TommyJackson85 CMS scopes are in progress in another PR: #1088.
> You can safely ignore that one for now, simply set the scopes to `[]` to let that widget always be visible, and we'll come back to it later.

Ok, no problem!

TommyJackson85 commented:
```pb:page:crud``` and ```forms:form:crud``` are now the scopes being used for their respective plugins and they appear to work as I've tested them manually. CMS Widgets remains accessable for all as its scopes are empty.
Are you happy to merge this PR?

Just for further context, from a previous issue I fixed, I added "settings" scopes for page builder, form builder and headless cms under settings in the side nav. It might be worth retesting them. Should the crud scopes be applied there too?

Pavel910 commented:
@TommyJackson85 Great! I think this is good for merging, and no you don’t need to add crud scopes to settings navigation items. Settings scopes are for settings, crud scopes are for crud modules. So it’s all good now. We’ll do further refinement later on, when the need arises.

I’ll merge this for the next release, feel free to change the draft status of PR and please update the PR body to remove the temporary text.

Thanks and good work! :)

TommyJackson85 commented:
How do I change the draft status? Will be good to learn this.
